### DISCUSSION TITLE
Possible Global Record Search Improvement

### DISCUSSION THREAD

**Discussion by seansan** [OpenMage/magento-lts#2112] [Ideas]
Possible Global Record Search Improvement

* If searching for full order increment_id *and* only 1 match: the redirect directly to order view page

* change order increment_id search to "%search_string" search from right because we often only enter the last XX digits

![image](https://user-images.githubusercontent.com/652395/46309818-d6522a80-c5bd-11e8-9ee3-56949a80807c.png)

**Comment by colinmollenhour** [comment]
Global search becomes totally unusable once you have a large order history. If anything it should only use prefix matches so that indexes may be used since when you use a suffix match a full table scan is required. With prefix matches or exact matches only the performance is pretty acceptable.

**Comment by seansan** [comment]
This is partly so. For us the solution was to re-think the usage. We dont use it to find customers (we can do that via orders) and we did not use it to find products. 

As a result we
- disabled customer and product search: it is only orders
- we reduced results set to 5
- added logic to the search query: number = order id
- only search for billing lastname (If text)
- add order_by (it is logic that it is often a recent order calling)
- Fix description to add the correct CustomerName + add the order date to the string
- Changed mathcing logic (lastname = always first letters, order id is always last numbers)
- Todo: implement when only 1 search result goto order redirect. UPDATE, found the trick: set `'autoSelect:true,'` in `new Ajax.Autocompleter('global_search',`

A lot is lost here. But also a lot is gained => because it is now being used again. And it is now very fast

My reasoning is. If nobody uses it. Then strip it to basics. And start adding logic that people *do use*!


Hope it helps someone

```
     * Load search results
     *
     * @return Mage_Adminhtml_Model_Search_Order
     */
    public function load()
    {
        $arr = array();

        if (!$this->hasStart() || !$this->hasLimit() || !$this->hasQuery()) {
            $this->setResults($arr);
            return $this;
        }

        $query = trim($this->getQuery());

        $search_array = array();
        if (is_numeric($query)) {
            $search_array[] = array('attribute' => 'increment_id', 'like'=>'%'.$query);
        } else {
            $search_array[] = array('attribute' => 'billing_lastname', 'like'=>$query.'%');
        }

        $collection = Mage::getResourceModel('sales/order_collection')
            ->addAttributeToSelect('*')
            ->addAttributeToSearchFilter($search_array)
            ->setCurPage($this->getStart())
            ->setPageSize($this->getLimit())
            ->setOrder('entity_id')
            ->load();

        foreach ($collection as $order) {
            $date = Mage::getModel('core/date')->date('d M Y', strtotime($order->getCreatedAt()));
            $arr[] = array(
                'id'                => 'order/1/'.$order->getId(),
                'type'              => Mage::helper('adminhtml')->__('Order'),
                'name'              => Mage::helper('adminhtml')->__("Order #%s", $order->getIncrementId()),
                'description'       => $order->getCustomerName() . " - $date",
                'form_panel_title'  => Mage::helper('adminhtml')->__('Order #%s (%s)', $order->getIncrementId(), $order->getBillingFirstname().' '.$order->getBillingLastname()),
                'url' => Mage::helper('adminhtml')->getUrl('*/sales_order/view', array('order_id'=>$order->getId())),
            );
        }

        $this->setResults($arr);

        return $this;
    }
```

![image](https://user-images.githubusercontent.com/652395/46333858-3a0e3f00-c622-11e8-959e-cd34e2250622.png)

**Comment by Flyingmana** [comment]
I would also say, if its becoming unusable with a large order history, thats a Bug worth to be fixed

**Comment by colinmollenhour** [comment]
The `is_numeric` optimization is a great one unless someone knows of a country where numbers are used in people's names.. 😆

**Comment by seansan** [comment]
Also thought of that

In Dutch we have a saying for this : looking for nails in shallow water

Translated something like: Listening for thunder when there is sunshine.

**Comment by colinmollenhour** [comment]
To be clear I was advocating *for* that optimization, not against it.

**Comment by seansan** [comment]
We have a saying for that too: thumbs up!
